### PR TITLE
Provide AIM support for Multi Layer Security Group

### PR DESCRIPTION
**Pull request by sanygupt**
This diff contains changes to add new resources to support operator
level security controls that are transparent to tenants.

### CODE REVIEW

**Comment by tbachman** [comment]
Looks like you have some formatting issues (see the PEP8 results from TravisCI)?

**Comment by tbachman** [comment]
Also - please change the title: since this is an open source project, we shouldn't reference Cisco-specific bug IDs.

**Comment by sanygupt** [comment]
> Looks like you have some formatting issues (see the PEP8 results from TravisCI)?

Fixed it now,

**Comment by sanygupt** [comment]
> Also - please change the title: since this is an open source project, we shouldn't reference Cisco-specific bug IDs.

Resolved.

File: aim/aim_manager.py [hunk 1]
```diff
@@ -177,6 +180,8 @@ def create(self, context, resource, overwrite=False, fix_ownership=False):
                     context.store.from_attr(old_db_obj, type(resource),
                                             attr_val)
             db_obj = old_db_obj or context.store.make_db_obj(resource)
+            if 'name' in dir(db_obj):
```
Comment (tbachman) [inline aim/aim_manager.py:183]: Isn't this done already in the resource's constructor object (which is needed in order to create the entry in the DB)?

File: aim/api/resource.py [hunk 2]
```diff
@@ -1751,3 +1751,103 @@ def __init__(self, **kwargs):
...
+        ('direction', t.enum("", "ingress", "egress")),
+        ('ethertype', t.enum("", "undefined", "ipv4", "ipv6")),
+        ('remote_ips', t.list_of_strings),
+        ('ip_protocol', t.string()),
+        ('from_port', t.port),
+        ('to_port', t.port),
+        ('conn_track', t.enum('normal', 'reflexive')),
+        ('icmp_type', t.string()),
+        ('icmp_code', t.string()),
+        ('remote_group_id', t.name),
```
Comment (tbachman) [inline aim/api/resource.py:1831]: I'm thinking we should get rid of this for hierarchical security groups - the workflow is for the operator to specify CIDRs that are allowed, not remote groups.

File: aim/db/migration/alembic_migrations/versions/5b00b6d805e9_system_security_groups.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        sa.Column('direction', sa.String(16)),
+        sa.Column('ip_protocol', sa.String(16)),
+        sa.Column('from_port', sa.String(16)),
+        sa.Column('to_port', sa.String(16)),
+        sa.Column('monitored', sa.Boolean, nullable=False, default=False),
+        sa.Column('epoch', sa.BigInteger(), nullable=False,
+                  server_default='0'),
+        sa.Column('conn_track', sa.String(25), nullable=False,
+                  server_default='reflexive'),
+        sa.Column('remote_group_id', sa.String(64),
```
Comment (tbachman) [inline aim/db/migration/alembic_migrations/versions/5b00b6d805e9_system_security_groups.py:85]: See above - we probably should remove remote_group_id.

File: aim/db/models.py [hunk 4]
```diff
@@ -1597,3 +1597,97 @@ class SpanSpanlbl(model_base.Base, model_base.HasDisplayName,
...
+                                  lazy='joined')
+    direction = sa.Column(sa.String(16))
+    ethertype = sa.Column(sa.String(16))
+    ip_protocol = sa.Column(sa.String(16))
+    from_port = sa.Column(sa.String(16))
+    to_port = sa.Column(sa.String(16))
+    conn_track = sa.Column(sa.String(25))
+    icmp_code = sa.Column(sa.String(16))
+    icmp_type = sa.Column(sa.String(16))
+    remote_group_id = sa.Column(sa.String(64), nullable=False)
```
Comment (tbachman) [inline aim/db/models.py:1666]: and again here - no remote_group_id.

**Review by tbachman — CHANGES_REQUESTED** [review]
Overall, a really clean patch! Just a few comments.

File: aim/aim_manager.py (hunk 1 above)
Comment (sanygupt) [reply aim/aim_manager.py:183]: This change was added by @sayalinaval . I think it would be better if she explains this change.

Comment (sayalinaval) [reply aim/aim_manager.py:183]: This can be removed.
And also line 245-246,273-274 should be removed from aim/tests/unit/test_aim_manager.py

Comment (sanygupt) [reply aim/aim_manager.py:183]: Thank you @sayalinaval , will do

**Review by tbachman — APPROVED** [review]

**Pull request merged**